### PR TITLE
Fix `with_halo` for `ImmersedBoundaryGrid` and clean up API

### LINKED ISSUE #2988: `with_halo` is broken for `ImmersedBoundaryGrid`
This code:

```julia
using Oceananigans
using Oceananigans.Grids: with_halo

underlying_grid = RectilinearGrid(size=(3, 3, 3), x=(0, 2), y=(0, 1), z=(0, 1))
grid = ImmersedBoundaryGrid(underlying_grid, GridFittedBottom((x, y) -> x))

with_halo((4, 4, 4), grid)
```

produces

```
ERROR: LoadError: ArgumentError: The dimensions of the immersed boundary (9, 9) do not match the grid size (11, 11)
Stacktrace:
 [1] validate_ib_size
   @ ~/.julia/packages/Oceananigans/QYJpb/src/ImmersedBoundaries/grid_fitted_immersed_boundaries.jl:92 [inlined]
 [2] ImmersedBoundaryGrid(grid::RectilinearGrid{Float64, Periodic, Periodic, Bounded, Float64, Float64, Float64, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, CPU}, ib::GridFittedBottom{OffsetArrays.OffsetMatrix{Float64, Matrix{Float64}}, Oceananigans.ImmersedBoundaries.CenterImmersedCondition})
   @ Oceananigans.ImmersedBoundaries ~/.julia/packages/Oceananigans/QYJpb/src/ImmersedBoundaries/grid_fitted_immersed_boundaries.jl:85
 [3] with_halo(halo::Tuple{Int64, Int64, Int64}, ibg::ImmersedBoundaryGrid{Float64, Periodic, Periodic, Bounded, RectilinearGrid{Float64, Periodic, Periodic, Bounded, Float64, Float64, Float64, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, CPU}, GridFittedBottom{OffsetArrays.OffsetMatrix{Float64, Matrix{Float64}}, Oceananigans.ImmersedBoundaries.CenterImmersedCondition}, Nothing, CPU})
   @ Oceananigans.ImmersedBoundaries ~/.julia/packages/Oceananigans/QYJpb/src/ImmersedBoundaries/ImmersedBoundaries.jl:140
 [4] top-level scope
   @ ~/Desktop/test.jl:9
 [5] include(fname::String)
   @ Base.MainInclude ./client.jl:476
 [6] top-level scope
   @ REPL[10]:1
 [7] top-level scope
   @ ~/.julia/packages/CUDA/BbliS/src/initialization.jl:52
```

This prevents us from writing `WENO(immersed_boundary_grid)`, since `with_halo` is used by `WENO` to calculate reconstruction coefficients for stretched grids.

### PR DESCRIPTION
**Pull request by glwagner**
This PR adds a docstring for `ImmersedBoundaryGrid` that includes the `active_cells_map` feature. It moves the interface to the top-level, and fixes `with_halo` to work with the new interface. We should add a test for `with_halo` with `ImmersedBoundaryGrid`.

resolves #2988

### CODE REVIEW

File: src/ImmersedBoundaries/ImmersedBoundaries.jl [hunk 1]
```diff
@@ -111,15 +111,36 @@ struct ImmersedBoundaryGrid{FT, TX, TY, TZ, G, I, M, Arch} <: AbstractGrid{FT, T
...
+        return new{FT, TX, TY, TZ, G, I, M, Arch}(arch, grid, ib, acm)
     end
 end
 
+"""
+    ImmersedBoundaryGrid(grid, immersed_boundary; active_cells_map=false)
+
+Return a grid with `immersed_boundary`. If `active_cells_map=true`,
+a linear index for every "active" (non-immersed) cell is generated, and
+the map between Carteian indices `(i, j, k)` and the active indices is
```
Comment (navidcy) [inline src/ImmersedBoundaries/ImmersedBoundaries.jl:128]:
```suggestion
the map between Cartesian indices `(i, j, k)` and the active indices is
```

File: src/ImmersedBoundaries/ImmersedBoundaries.jl [hunk 2]
```diff
@@ -137,7 +158,16 @@ const IBG = ImmersedBoundaryGrid
 Adapt.adapt_structure(to, ibg::IBG{FT, TX, TY, TZ}) where {FT, TX, TY, TZ} =
     ImmersedBoundaryGrid{TX, TY, TZ}(adapt(to, ibg.underlying_grid), adapt(to, ibg.immersed_boundary), adapt(to, ibg.active_cells_map))
 
-with_halo(halo, ibg::ImmersedBoundaryGrid) = ImmersedBoundaryGrid(with_halo(halo, ibg.underlying_grid), ibg.immersed_boundary)
+function with_halo(halo, ibg::ImmersedBoundaryGrid)
+    TX, TY, TZ = topology(ibg)
+
+    # I think we have to recompute the active cells map?
+    active_cells_map = !isnothing(ibg.active_cells_map)
```
Comment (navidcy) [inline src/ImmersedBoundaries/ImmersedBoundaries.jl:165]: hm....
looking here --> https://github.com/CliMA/Oceananigans.jl/blob/19dbe19a76043100ac7fd08bdd754cc362fc19dd/src/ImmersedBoundaries/active_cells_map.jl#L29

reads to me that the active_cells_map is only about the interior? That means we don't need to recompute the active cells map? is this true? remove comment if so?

**Review by navidcy — APPROVED** [review]

File: src/ImmersedBoundaries/ImmersedBoundaries.jl (hunk 2 above)
Comment (simone-silvestri) [reply src/ImmersedBoundaries/ImmersedBoundaries.jl:165]: this is indeed true

**Comment by glwagner** [comment]
@simone-silvestri since you are superceding this PR, can you please take into account some the API improvements as well?

**Pull request closed without merging**